### DISCUSSION TITLE
Properties of embedded images

### DISCUSSION THREAD

**Discussion by pglizniewicz** [sumatrapdfreader/sumatrapdf#3553] [Q&A]
How can I check the properties of images embedded in PDF files? I know I can copy each page as an image, but I haven't found and option to just check its properties (especially compression parameters).

I often read scanned books, basically one image for one page and some OCRed text. Some of these work great, they render instantly, others not so much, I often have to wait several seconds to render the page. I would like to know, how they differ to "optimize" the slow PDFs. I've tried dozens of PDF optimizers, they would often shrink the PDF's size, but none of them make them render faster.

**Comment by GitHubRulesOK** [comment]
Hmm very good question, there are several dozens of ways an image can be stored in a PDF, but it usually boils down to 2 major types plus exotics. the one I have been working on last says `Image/Width 37/Height 37/BitsPerComponent 1/ColorSpace/DeviceGray/Length 185` at that size its clearly a QRcoded mono bitmap, so without compression its instant to view. If I compress it it will be oddly bigger (often the case with smaller images) and slower to decompress `/Image/Width 37/Height 37/BitsPerComponent 1/ColorSpace/DeviceGray/Length 294/Filter/CCITTFaxDecode>>`

1) JPEG these are normally stored untouched thus have all those characteristics such as poor definition poor use of colour because of the blocky compression. They cannot usually be reduced in size, because the compression is already optimal for that lossy content, thus "compressors" need to degrade the quality more by removing 3 out of 4 pixels (retaining say 25% on each pass)

2) High quality lossless images usually stored without their headers that previously tell you what type they were. These are usually optimally compressed as de"Flated" ideally TiFF Monochrome will be High Speed CITTFAX as a good balance between decompression and speed of reading. PNG will usually be stored as two overlay bitmap images thus both need to be read at same time, so a bit slower but pixel perfect as each zoom pass is carried out to the point of full display. Again its not normally possible to improve the compression already used, without stripping out good contents.

3) Exotics like WebP (NOT found in PDF as too new) so proprietary JBIG2k etc highly compressed thus slow to decompress, these could sometimes benefit by making them bigger, thus faster to decompress and render.

Basically it depends how badly the document was constructed as to what degrading may help. usually its not the image slow down a scanned source (some are highly optimised for speed) but the massive bloat that OCR itself uses/introduces.


Konica Scanners use a special "compact" mode where faster masks are seen above slower jpeg background so its potentially faster to see the rendered high res text as a "stencil" before the glossy lower resolution fast backdrop.

**Comment by pglizniewicz** [comment]
Thanks for replying. If dug deeper, and this is what I saw.

The "fast rendering" book uses JPEG for its content. This is the output of pdfimages (don't know any alternative tool to check this) for the first 10 pages is:
```
page   num  type   width height color comp bpc  enc interp  object ID x-ppi y-ppi size ratio
--------------------------------------------------------------------------------------------
   1     0 image    4065  4547  icc     3   8  jpeg   no      2941  0    72    72 4491K 8.3%
   2     1 image    5100  6600  gray    1   8  jpeg   no        10  0   600   600 1028K 3.1%
   2     2 stencil     1     1  -       1   1  image  no   [inline]   0.196     4    0B   -
   3     3 image    5190  6671  gray    1   8  jpeg   no        14  0   600   600  138K 0.4%
   4     4 image    5113  6609  rgb     3   8  jpeg   no        18  0   600   600 6671K 6.7%
   4     5 stencil     1     1  -       1   1  image  no   [inline]   0.195     4    0B   -
   4     6 stencil     1     1  -       1   1  image  no   [inline]       2     3    0B   -
   4     7 stencil     1     1  -       1   1  image  no   [inline]   0.915     5    0B   -
   4     8 stencil     1     1  -       1   1  image  no   [inline]      25 0.694    0B   -
   4     9 stencil     1     1  -       1   1  image  no   [inline]      25 0.322    0B   -
   4    10 stencil     1     1  -       1   1  image  no   [inline]   0.789     2    0B   -
   4    11 stencil     1     1  -       1   1  image  no   [inline]   0.118     2    0B   -
   5    12 image    5357  6810  gray    1   8  jpeg   no        22  0   600   600  145K 0.4%
   6    13 image    5113  6609  gray    1   8  jpeg   no        26  0   600   600 1686K 5.1%
   6    14 stencil     1     1  -       1   1  image  no   [inline]   0.196     4    0B   -
   7    15 image    5357  6810  gray    1   8  jpeg   no        30  0   600   600  144K 0.4%
   8    16 image    5100  6600  rgb     3   8  jpeg   no        34  0   600   600 7954K 8.1%
   8    17 stencil     1     1  -       1   1  image  no   [inline]   0.167     2    0B   -
   8    18 stencil     1     1  -       1   1  image  no   [inline]   0.528     3    0B   -
   9    19 image    5357  6810  gray    1   8  jpeg   no        38  0   600   600  145K 0.4%
  10    20 image    5126  6620  gray    1   8  jpeg   no        42  0   600   600  916K 2.8%
  10    21 stencil     1     1  -       1   1  image  no   [inline]   0.196     4    0B   -
```

Now the slow book seems to be using much lower resolution images, which I somehow thought are the cause of slow rendering, but are not the problem. I see compression `jpx`, which as far as I understand is JPEG 2000:

```
page   num  type   width height color comp bpc  enc interp  object ID x-ppi y-ppi size ratio
--------------------------------------------------------------------------------------------
   1     0 image    1190  1491  rgb     3   8  jpx    no     11323  0   100   100 34.6K 0.7%
   1     1 image    3569  4473  rgb     3   8  jpx    no     11324  0   300   300 79.5K 0.2%
   1     2 mask     3569  4473  -       1   1  jpx    no     11324  0   300   300 79.5K 4.1%
   2     3 image    1144  1476  rgb     3   8  jpx    no         4  0   100   100 6354B 0.1%
   2     4 image    3432  4427  rgb     3   8  jpx    no         5  0   300   300 19.0K 0.0%
   2     5 mask     3432  4427  -       1   1  jpx    no         5  0   300   300 19.0K 1.0%
   3     6 image    1159  1479  rgb     3   8  jpx    no        10  0   100   100 7263B 0.1%
   3     7 image    3477  4436  rgb     3   8  jpx    no        11  0   300   300 53.6K 0.1%
   3     8 mask     3477  4436  -       1   1  jpx    no        11  0   300   300 53.6K 2.8%
   4     9 image    1144  1476  rgb     3   8  jpx    no        16  0   100   100 6732B 0.1%
   4    10 image    3432  4427  rgb     3   8  jpx    no        17  0   300   300 51.4K 0.1%
```

If JPEG2000 compression is the problem, then maybe I can just "recompress" embedded images to JPEG? Or any other "faster" format? I've looked at some utilities, and why I know I can extract images with pdfimages from poppler and I know how to convert them with ImageMagic or any other batch images processor, I can't find any utility that would "replace" the pages without loosing OCR (which I use mostly for searching inside the PDF).

**Comment by GitHubRulesOK** [comment]
hmmm
 Well done for research and that is an odd mix but
faster is more like 600 dpi conventional jpg with odd artefacts ( 1 x 1 pixel images ? ) thus reduction to 300 dpi should help but depends on quality
Slower is odd 3 images overlaid per page (3 times to render) 2 at higher density suggests May have been adapted PNG Alpha but why a third perhaps only needs 1 of the 3 and such high object number?  cover page added as last or set Linearized. ?
```
page   num  type   width height color comp bpc  enc interp  object ID x-ppi y-ppi size ratio
--------------------------------------------------------------------------------------------
   1     0 image    1190  1491  rgb     3   8  jpx    no     11323  0   100   100 34.6K 0.7%
   1     1 image    3569  4473  rgb     3   8  jpx    no     11324  0   300   300 79.5K 0.2%
   1     2 mask     3569  4473  -       1   1  jpx    no     11324  0   300   300 79.5K 4.1%
```
Reprint as one image would loose OCR data so would need some selective process to decide what stays and goes, So what springs to mind is GhostScript can alter formats, but unsure how, it would need to be directed, as not designed to crop out components except via PostScripting. 
To selectively remove objects would probably need qpdf or mutool (or similar) decompression then scripted delete from the list of objects but again thats a tough task  without clear directions

**Comment by pglizniewicz** [comment]
Okay, I've tried a simple `gs -sDEVICE=pdfwrite -o output.pdf input.pdf` on the JPX compressed, the file grew from 45 MB to 226 MB. It's blazing fast in SumatraPDF, pdfimages shows the big images are now JPG, at least the big ones, the ones at 1/3 resolution are still JPX. Adding `-dPDFSETTINGS=/ebook` reduces DPI to 150, what is still OK for me, and the size shrinks to 114 MB. Guess I will have to experiment with some other quality settings to get the PDF to reasonable file size.

**Reply by GitHubRulesOK** [reply]
hmm
you seem to be on the right tack **bigger can be faster to display**
as you are finding there is a point of balance where each documents quality can be reduced yet maintain readability at speed.

default /screen res is 96dpi which for some sources (say a screenshot) can be bitmap perfect but actually be degraded by using a higher density like 100 dpi which itself for the same given page area needs more compression thus more delay in unpacking !
